internal/tests/helper/testing_helper.go

```diff
@@ -107,34 +108,67 @@ func NewTestSession(
 func (s *TestSession) connect(ctx context.Context, t *testing.T) net.Conn {
 	t.Helper()
 	require := require.New(t)
-	conn, resp, err := websocket.Dial(
-		ctx,
-		fmt.Sprintf("wss://%s/v1/proxy", s.WorkerAddr),
-		&websocket.DialOptions{
-			HTTPClient: &http.Client{
-				Transport: s.Transport,
-			},
-			Subprotocols: []string{globals.TcpProxyV1},
-		},
-	)
-	require.NoError(err)
-	require.NotNil(conn)
-	require.NotNil(resp)
-	require.Equal(resp.Header.Get("Sec-WebSocket-Protocol"), globals.TcpProxyV1)
 
-	// Send the handshake.
-	if s.tofuToken == "" {
-		s.tofuToken, err = base62.Random(20)
-	}
+	var conn *websocket.Conn
+	var resp *http.Response
+	var err error
+	var handshakeResult proxy.HandshakeResult
+	// A retry was added here to mitigate some flakiness.
+	// Occasionally, `wspb.Read` would throw an error due to "received header with unexpected
+	// rsv bits set: false:false:true". It was unclear what the cause of this was, so we
+	// resorted to retrying the connection.
+	err = backoff.RetryNotify(
+		func() error {
+			conn, resp, err = websocket.Dial(
+				ctx,
+				fmt.Sprintf("wss://%s/v1/proxy", s.WorkerAddr),
+				&websocket.DialOptions{
+					HTTPClient: &http.Client{
+						Transport: s.Transport,
+					},
+					Subprotocols: []string{globals.TcpProxyV1},
+				},
+			)
```

Comment:
You can declare the error here to limit the scope.
```suggestion
			var err error
			conn, resp, err = websocket.Dial(
				ctx,
				fmt.Sprintf("wss://%s/v1/proxy", s.WorkerAddr),
				&websocket.DialOptions{
					HTTPClient: &http.Client{
						Transport: s.Transport,
					},
					Subprotocols: []string{globals.TcpProxyV1},
				},
			)
```